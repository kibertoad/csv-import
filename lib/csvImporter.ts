import { Options, parse } from 'csv-parse'
import { Readable } from 'stream'
import { Mapper } from './mappers'
import { ValidationContext, ValidationResult } from './validation'

export type ColumnDefinition<T> = {
  column: string
  entityField: string
  mapper: Mapper<T>
  defaultValue?: T
}

export type ColumnDefinitions = ColumnDefinition<any>[]

export type ImportResult<T> = {
  data: T[]
  validationResult: ValidationResult
  importTime: number
}

export function loadFromCsv<T extends Record<string, any>>(
  stream: Readable | Buffer | string,
  columnConfig: ColumnDefinitions,
  csvOptions: Omit<Options, 'columns'>
): Promise<ImportResult<T>> {
  return new Promise((resolve, reject) => {
    const startTime = new Date()
    const validationContext = new ValidationContext({
      fromLine: csvOptions.fromLine || csvOptions.from_line,
    })
    const result: T[] = []
    const columns = columnConfig.map((column) => column.column)
    const parser =
      stream instanceof Readable
        ? parse({
            ...csvOptions,
            columns,
          })
        : parse(stream, {
            ...csvOptions,
            columns,
          })

    parser.on('readable', () => {
      let row
      while ((row = parser.read()) !== null) {
        const entity: Record<string, any> = {}
        for (var i = 0; i < columnConfig.length; i++) {
          const columnConfigEntry = columnConfig[i]
          entity[columnConfigEntry.entityField] = columnConfigEntry.mapper(
            columnConfigEntry.column,
            row[columnConfigEntry.column],
            validationContext,
            columnConfigEntry.defaultValue ?? undefined
          )
        }
        result.push(entity as T)
        validationContext.nextRow()
      }
    })
    parser.on('error', (err) => {
      reject(err)
    })
    parser.on('end', () => {
      const endDate = new Date()
      resolve({
        data: result,
        validationResult: validationContext.getResult(),
        importTime: endDate.getTime() - startTime.getTime(),
      })
    })

    if (stream instanceof Readable) {
      stream.pipe(parser)
    }
  })
}
