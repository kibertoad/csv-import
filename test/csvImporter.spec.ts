import { resolve } from 'path'
import { getFileStream } from './utils/fileUtils'
import { ColumnDefinitions, loadFromCsv } from '../lib/csvImporter'
import {
  toMandatoryDate,
  toMandatoryFloat,
  toMandatoryInteger,
  toMandatoryString,
  toOptionalBoolean,
  toOptionalDate,
  toOptionalFloat,
  toOptionalInteger,
  toOptionalString,
} from '../lib/mappers'

describe('csvImporter', () => {
  describe('loadEntities', () => {
    it('correctly loads large file', async () => {
      const filePath = resolve(__dirname, 'data', 'stop_times.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'trip_id',
          entityField: 'tripId',
          mapper: toMandatoryInteger,
        },
        {
          column: 'arrival_time',
          entityField: 'arrival_time',
          mapper: toOptionalString,
        },
        {
          column: 'departure_time',
          entityField: 'departure_time',
          mapper: toMandatoryString,
        },
        {
          column: 'stop_id',
          entityField: 'stop_id',
          mapper: toMandatoryInteger,
        },
        {
          column: 'stop_sequence',
          entityField: 'stop_sequence',
          mapper: toMandatoryInteger,
        },
        {
          column: 'pickup_type',
          entityField: 'pickup_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'drop_off_type',
          entityField: 'drop_off_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'destinationTraveled',
          entityField: 'destinationTraveled',
          mapper: toOptionalInteger,
          defaultValue: 0,
        },
        {
          column: 'priority',
          entityField: 'priority',
          mapper: toOptionalBoolean,
          defaultValue: false,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(512781)
      expect(result.data[0]).toMatchSnapshot()
      expect(result.data[8785]).toMatchSnapshot()
      expect(result.data[112780]).toMatchSnapshot()
      expect(result.data[212780]).toMatchSnapshot()
      expect(result.data[312780]).toMatchSnapshot()
      expect(result.data[512760]).toMatchSnapshot()
      expect(result.data[512780]).toMatchSnapshot()
    }, 90000)

    it('correctly loads medium file', async () => {
      const filePath = resolve(__dirname, 'data', 'stop_times_mid.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'trip_id',
          entityField: 'tripId',
          mapper: toMandatoryInteger,
        },
        {
          column: 'arrival_time',
          entityField: 'arrival_time',
          mapper: toOptionalString,
        },
        {
          column: 'departure_time',
          entityField: 'departure_time',
          mapper: toMandatoryString,
        },
        {
          column: 'stop_id',
          entityField: 'stop_id',
          mapper: toMandatoryInteger,
        },
        {
          column: 'stop_sequence',
          entityField: 'stop_sequence',
          mapper: toMandatoryInteger,
        },
        {
          column: 'pickup_type',
          entityField: 'pickup_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'drop_off_type',
          entityField: 'drop_off_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'destinationTraveled',
          entityField: 'destinationTraveled',
          mapper: toOptionalFloat,
          defaultValue: 0.0,
        },
        {
          column: 'priority',
          entityField: 'priority',
          mapper: toOptionalBoolean,
          defaultValue: true,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.data.length).toEqual(5040)
      expect(result.data[0]).toMatchSnapshot()
      expect(result.data[1020]).toMatchSnapshot()
      expect(result.data[5030]).toMatchSnapshot()
      expect(result.data[5039]).toMatchSnapshot()
    }, 90000)

    it('correctly returns errors', async () => {
      const filePath = resolve(__dirname, 'data', 'stop_times_errors.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'trip_id',
          entityField: 'tripId',
          mapper: toMandatoryString,
        },
        {
          column: 'arrival_time',
          entityField: 'arrival_time',
          mapper: toOptionalDate,
        },
        {
          column: 'departure_time',
          entityField: 'departure_time',
          mapper: toMandatoryDate,
        },
        {
          column: 'stop_id',
          entityField: 'stop_id',
          mapper: toMandatoryInteger,
        },
        {
          column: 'stop_sequence',
          entityField: 'stop_sequence',
          mapper: toMandatoryInteger,
        },
        {
          column: 'pickup_type',
          entityField: 'pickup_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'drop_off_type',
          entityField: 'drop_off_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'destinationTraveled',
          entityField: 'destinationTraveled',
          mapper: toOptionalString,
        },
        {
          column: 'priority',
          entityField: 'priority',
          mapper: toOptionalBoolean,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(7)
      expect(result.data).toMatchSnapshot()
    })

    it('correctly loads string stream', async () => {
      const input =
        'vehicleGarageNumber,longitude,latitude,speed,azimuth,tripStartInMinutes,deflectionInSeconds,time,vehicleType,routeShortName\n7015,25.20696,54.64424,0,260,1314,0,2021-12-13 19:30:28,bus,51\n8048,25.17336,54.70325,37,210,1315,0,2021-12-13 19:30:28,bus,4G\n7044,25.20286,54.68304,30,246,1339,5,2021-12-13 19:30:28,bus,116\n3039,25.29422,54.67497,38,80,1347,114,2021-12-13 19:30:28,bus,31\n2691,25.22013,54.69601,46,24,1329,-66,2021-12-13 19:30:28,trolleybus,16\n'

      const columnConfig: ColumnDefinitions = [
        {
          column: 'vehicleGarageNumber',
          entityField: 'vehicleGarageNumber',
          mapper: toMandatoryString,
        },
        {
          column: 'longitude',
          entityField: 'longitude',
          mapper: toMandatoryFloat,
        },
        {
          column: 'latitude',
          entityField: 'latitude',
          mapper: toMandatoryFloat,
        },
        {
          column: 'speed',
          entityField: 'speed',
          mapper: toMandatoryInteger,
        },
        {
          column: 'azimuth',
          entityField: 'azimuth',
          mapper: toMandatoryInteger,
        },
        {
          column: 'tripStartInMinutes',
          entityField: 'tripStartInMinutes',
          mapper: toOptionalInteger,
        },
        {
          column: 'deflectionInSeconds',
          entityField: 'deflectionInSeconds',
          mapper: toOptionalInteger,
        },
        {
          column: 'time',
          entityField: 'time',
          mapper: toMandatoryDate,
        },
        {
          column: 'vehicleTypeName',
          entityField: 'vehicleTypeName',
          mapper: toMandatoryString,
        },

        {
          column: 'routeShortName',
          entityField: 'routeShortName',
          mapper: toMandatoryString,
        },
      ]
      const result = await loadFromCsv(input, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(5)
      expect(result.data).toMatchSnapshot()
    })

    it('correctly returns string stream errors', async () => {
      const input =
        'vehicleGarageNumber,longitude,latitude,speed,azimuth,tripStartInMinutes,deflectionInSeconds,time,vehicleType,routeShortName\n7015,long,54.64424,0,260,1314,0,2021-12-13 19:30:28,bus,51\n8048,25.17336,54.70325,37,210,1315,0,datt,bus,4G\n7044,25.20286,54.68304,30,246,1339,5,2021-12-13 19:30:28,bus,116\n3039,25.29422,54.67497,speed,80,1347,114,2021-12-13 19:30:28,bus,31\n2691,25.22013,54.69601,46,24,1329,-66,2021-12-13 19:30:28,trolleybus,16\n'

      const columnConfig: ColumnDefinitions = [
        {
          column: 'vehicleGarageNumber',
          entityField: 'vehicleGarageNumber',
          mapper: toMandatoryString,
        },
        {
          column: 'longitude',
          entityField: 'longitude',
          mapper: toMandatoryFloat,
        },
        {
          column: 'latitude',
          entityField: 'latitude',
          mapper: toMandatoryFloat,
        },
        {
          column: 'speed',
          entityField: 'speed',
          mapper: toMandatoryInteger,
        },
        {
          column: 'azimuth',
          entityField: 'azimuth',
          mapper: toMandatoryInteger,
        },
        {
          column: 'tripStartInMinutes',
          entityField: 'tripStartInMinutes',
          mapper: toOptionalInteger,
        },
        {
          column: 'deflectionInSeconds',
          entityField: 'deflectionInSeconds',
          mapper: toOptionalInteger,
        },
        {
          column: 'time',
          entityField: 'time',
          mapper: toMandatoryDate,
        },
        {
          column: 'vehicleTypeName',
          entityField: 'vehicleTypeName',
          mapper: toMandatoryString,
        },

        {
          column: 'routeShortName',
          entityField: 'routeShortName',
          mapper: toMandatoryString,
        },
      ]
      const result = await loadFromCsv(input, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(5)
      expect(result.data).toMatchSnapshot()
    })
  })
})
