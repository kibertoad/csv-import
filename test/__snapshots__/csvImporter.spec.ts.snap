// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvImporter loadEntities correctly loads large file 1`] = `
Object {
  "errors": Array [],
  "isValid": true,
}
`;

exports[`csvImporter loadEntities correctly loads large file 2`] = `
Object {
  "arrival_time": "04:49:00",
  "departure_time": "04:50:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 8262,
  "stop_sequence": 1,
  "tripId": 1,
}
`;

exports[`csvImporter loadEntities correctly loads large file 3`] = `
Object {
  "arrival_time": "15:23:00",
  "departure_time": "15:23:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 5836,
  "stop_sequence": 6,
  "tripId": 716,
}
`;

exports[`csvImporter loadEntities correctly loads large file 4`] = `
Object {
  "arrival_time": "11:38:00",
  "departure_time": "11:38:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 6083,
  "stop_sequence": 9,
  "tripId": 6375,
}
`;

exports[`csvImporter loadEntities correctly loads large file 5`] = `
Object {
  "arrival_time": "11:22:00",
  "departure_time": "11:22:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 16581,
  "stop_sequence": 12,
  "tripId": 11523,
}
`;

exports[`csvImporter loadEntities correctly loads large file 6`] = `
Object {
  "arrival_time": "22:26:00",
  "departure_time": "22:26:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 6497,
  "stop_sequence": 31,
  "tripId": 16273,
}
`;

exports[`csvImporter loadEntities correctly loads large file 7`] = `
Object {
  "arrival_time": "19:22:00",
  "departure_time": "19:22:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 5086,
  "stop_sequence": 3,
  "tripId": 25564,
}
`;

exports[`csvImporter loadEntities correctly loads large file 8`] = `
Object {
  "arrival_time": "06:14:00",
  "departure_time": "06:14:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 5505,
  "stop_sequence": 8,
  "tripId": 25566,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 1`] = `
Object {
  "arrival_time": "04:49:00",
  "departure_time": "04:50:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 8262,
  "stop_sequence": 1,
  "tripId": 1,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 2`] = `
Object {
  "arrival_time": "07:26:00",
  "departure_time": "07:26:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": true,
  "stop_id": 7573,
  "stop_sequence": 8,
  "tripId": 127,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 3`] = `
Object {
  "arrival_time": "13:22:00",
  "departure_time": "13:22:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": true,
  "stop_id": 5156,
  "stop_sequence": 1,
  "tripId": 466,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 4`] = `
Object {
  "arrival_time": "13:49:00",
  "departure_time": "13:49:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": true,
  "stop_id": 6487,
  "stop_sequence": 10,
  "tripId": 466,
}
`;

exports[`csvImporter loadEntities correctly loads string stream 1`] = `
Object {
  "errors": Array [],
  "isValid": true,
}
`;

exports[`csvImporter loadEntities correctly loads string stream 2`] = `
Array [
  Object {
    "azimuth": 260,
    "deflectionInSeconds": 0,
    "latitude": 54.64424,
    "longitude": 25.20696,
    "routeShortName": "51",
    "speed": 0,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1314,
    "vehicleGarageNumber": "7015",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 210,
    "deflectionInSeconds": 0,
    "latitude": 54.70325,
    "longitude": 25.17336,
    "routeShortName": "4G",
    "speed": 37,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1315,
    "vehicleGarageNumber": "8048",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 246,
    "deflectionInSeconds": 5,
    "latitude": 54.68304,
    "longitude": 25.20286,
    "routeShortName": "116",
    "speed": 30,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1339,
    "vehicleGarageNumber": "7044",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 80,
    "deflectionInSeconds": 114,
    "latitude": 54.67497,
    "longitude": 25.29422,
    "routeShortName": "31",
    "speed": 38,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1347,
    "vehicleGarageNumber": "3039",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 24,
    "deflectionInSeconds": -66,
    "latitude": 54.69601,
    "longitude": 25.22013,
    "routeShortName": "16",
    "speed": 46,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1329,
    "vehicleGarageNumber": "2691",
    "vehicleTypeName": "trolleybus",
  },
]
`;

exports[`csvImporter loadEntities correctly returns errors 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "arrival_time",
      "line": 3,
    },
    Object {
      "code": "INVALID_INTEGER_FORMAT",
      "column": "pickup_type",
      "line": 3,
    },
    Object {
      "code": "MISSING_MANDATORY_FIELD",
      "column": "drop_off_type",
      "line": 4,
    },
    Object {
      "code": "INVALID_BOOLEAN_FORMAT",
      "column": "priority",
      "line": 4,
    },
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "arrival_time",
      "line": 5,
    },
    Object {
      "code": "MISSING_MANDATORY_FIELD",
      "column": "departure_time",
      "line": 6,
    },
    Object {
      "code": "MISSING_MANDATORY_FIELD",
      "column": "trip_id",
      "line": 8,
    },
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "departure_time",
      "line": 8,
    },
  ],
  "isValid": false,
}
`;

exports[`csvImporter loadEntities correctly returns errors 2`] = `
Array [
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": false,
    "stop_id": 8262,
    "stop_sequence": 1,
    "tripId": "a",
  },
  Object {
    "arrival_time": undefined,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": undefined,
    "priority": undefined,
    "stop_id": 6650,
    "stop_sequence": 2,
    "tripId": "b",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": undefined,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 7851,
    "stop_sequence": 3,
    "tripId": "c",
  },
  Object {
    "arrival_time": undefined,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 5173,
    "stop_sequence": 4,
    "tripId": "d",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": undefined,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": true,
    "stop_id": 5175,
    "stop_sequence": 5,
    "tripId": "e",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 10359,
    "stop_sequence": 6,
    "tripId": "f",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": undefined,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 9738,
    "stop_sequence": 7,
    "tripId": undefined,
  },
]
`;

exports[`csvImporter loadEntities correctly returns string stream errors 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "INVALID_FLOAT_FORMAT",
      "column": "longitude",
      "line": 2,
    },
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "time",
      "line": 3,
    },
    Object {
      "code": "INVALID_INTEGER_FORMAT",
      "column": "speed",
      "line": 5,
    },
  ],
  "isValid": false,
}
`;

exports[`csvImporter loadEntities correctly returns string stream errors 2`] = `
Array [
  Object {
    "azimuth": 260,
    "deflectionInSeconds": 0,
    "latitude": 54.64424,
    "longitude": undefined,
    "routeShortName": "51",
    "speed": 0,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1314,
    "vehicleGarageNumber": "7015",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 210,
    "deflectionInSeconds": 0,
    "latitude": 54.70325,
    "longitude": 25.17336,
    "routeShortName": "4G",
    "speed": 37,
    "time": undefined,
    "tripStartInMinutes": 1315,
    "vehicleGarageNumber": "8048",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 246,
    "deflectionInSeconds": 5,
    "latitude": 54.68304,
    "longitude": 25.20286,
    "routeShortName": "116",
    "speed": 30,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1339,
    "vehicleGarageNumber": "7044",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 80,
    "deflectionInSeconds": 114,
    "latitude": 54.67497,
    "longitude": 25.29422,
    "routeShortName": "31",
    "speed": undefined,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1347,
    "vehicleGarageNumber": "3039",
    "vehicleTypeName": "bus",
  },
  Object {
    "azimuth": 24,
    "deflectionInSeconds": -66,
    "latitude": 54.69601,
    "longitude": 25.22013,
    "routeShortName": "16",
    "speed": 46,
    "time": 2021-12-13T19:30:28.000Z,
    "tripStartInMinutes": 1329,
    "vehicleGarageNumber": "2691",
    "vehicleTypeName": "trolleybus",
  },
]
`;
